### PR TITLE
Fix Helm chart does not support special characters in access/secret key

### LINKED ISSUE #15221: Helm chart does not support special characters in access/secret key
You can have the Helm chart in this repository create users in Minio by specifying values like this:

```yaml
users:
    - accessKey: xxx
      secretKey: xxx
      policy: consoleAdmin
```

This works fine as long as the `accessKey` and `secretKey` don't contain any special characters. However, when I tried the following values:

```yaml
users:
  - accessKey: "tv:(()KW5$`>@/c:dB6_=.4|{jv?@!"
    secretKey: ".B_ES205}Mjzrh}G{emJi'LEuUE#:R"
    policy: consoleAdmin
```

The `minio-make-user-job` pod failed with the following error:

```
/config/add-user: line 72: syntax error near unexpected token `('
```

## Expected Behavior
I expected the chart to create a user with the specified access key and secret key.

## Current Behavior
The chart fails to create the user.

## Possible Solution
The problem is that the access key and secret key are inserted verbatim into a bash script:

https://github.com/minio/minio/blob/ce667ddae007a13df0020f6879fe5e0f955ba368/helm/minio/templates/_helper_create_user.txt#L78-L88

That means that characters like ` (backtick) and ( (parenthesis) are interpreted as bash syntax rather than a literal value. The script should properly escape these values, for example by encoding them using base64.

## Steps to Reproduce (for bugs)
Deploy the chart with the values mentioned above.

## Context
I'm trying to create a user with a secure password.

## Regression
As far as I can tell this has always been broken.

## Your Environment

* Version used (`minio --version`): N/A
* Server setup and configuration: N/A
* Operating System and version (`uname -a`): N/A

### PR DESCRIPTION
**Pull request by dormanze**
## Description
fix #15221 

## Motivation and Context
Fix Helm chart does not support special characters in access/secret key.

## How to test this PR?
prefabricate users with access/secretkey containing special characters.
like:
```
users:
  - accessKey: "tv:(()KW5$`>@/c:dB6_=.4|{jv?@!"
    secretKey: ".B_ES205}Mjzrh}G{emJi'LEuUE#:R"
    policy: consoleAdmin
```
image

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by harshavardhana** [comment]
@dormanze do not see any files changed ?

**Comment by dormanze** [comment]
> @dormanze do not see any files changed ?
Sorry, there is something wrong with my Internet.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
{"detail": "Not found"}Deleting image locally

**Review by harshavardhana — CHANGES_REQUESTED** [review]
This change is too complicated what is the issue with helm chart not supporting these characters?

**Comment by dormanze** [comment]
> This change is too complicated what is the issue with helm chart not supporting these characters?
Thank you for your attention.
It has nothing to do with helm.
In the shell, characters like {,',< will be parsed as keywords, so it will cause error messages.
You can create users by executing the following commands in the shell.
```
#accessKey is : tv:(()KW5$`>@/c:dB6_=.4|{jv?@!
#secretKey is : .B_ES205}Mjzrh}G{emJi'LEuUE#:R
mc admin user add myminio tv:(()KW5$`>@/c:dB6_=.4|{jv?@! .B_ES205}Mjzrh}G{emJi'LEuUE#:R --insecure
```
image

**Comment by harshavardhana** [comment]
> #accessKey is : tv:(()KW5$`>@/c:dB6_=.4|{jv?@!
> #secretKey is : .B_ES205}Mjzrh}G{emJi'LEuUE#:R
> mc admin user add myminio tv:(()KW5$`>@/c:dB6_=.4|{jv?@! .B_ES205}Mjzrh}G{emJi'LEuUE#:R --insecure

All we need to do is save both the credentials in a file

```
tv:(()KW5$`>@/c:dB6_=.4|{jv?@!
.B_ES205}Mjzrh}G{emJi'LEuUE#:R
```

and then

```
cat credentials.txt | mc admin user add alias/
```

The PR should be simplified and not add new confusing variables.

**Comment by dormanze** [comment]
> 



> > #accessKey is : tv:(()KW5$`>@/c:dB6_=.4|{jv?@! #secretKey is : .B_ES205}Mjzrh}G{emJi'LEuUE#:R mc admin user add myminio tv:(()KW5$`>@/c:dB6_=.4|{jv?@! .B_ES205}Mjzrh}G{emJi'LEuUE#:R --insecure
> 
> All we need to do is save both the credentials in a file
> 
> ```
> tv:(()KW5$`>@/c:dB6_=.4|{jv?@!
> .B_ES205}Mjzrh}G{emJi'LEuUE#:R
> ```
> 
> and then
> 
> ```
> cat credentials.txt | mc admin user add alias/
> ```
> 
> The PR should be simplified and not add new confusing variables.



> > #accessKey is : tv:(()KW5$`>@/c:dB6_=.4|{jv?@! #secretKey is : .B_ES205}Mjzrh}G{emJi'LEuUE#:R mc admin user add myminio tv:(()KW5$`>@/c:dB6_=.4|{jv?@! .B_ES205}Mjzrh}G{emJi'LEuUE#:R --insecure
> 
> All we need to do is save both the credentials in a file
> 
> ```
> tv:(()KW5$`>@/c:dB6_=.4|{jv?@!
> .B_ES205}Mjzrh}G{emJi'LEuUE#:R
> ```
> 
> and then
> 
> ```
> cat credentials.txt | mc admin user add alias/
> ```
> 
> The PR should be simplified and not add new confusing variables.

This is really a good idea. I'll revise my pr.

**Comment by dormanze** [comment]
@harshavardhana Please review again.thank you very much.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: helm/minio/templates/_helper_create_user.txt [hunk 1]
```diff
@@ -35,28 +39,35 @@ connectToMinio() {
...
-  POLICY=$3
-
+  POLICY=$1
+  #check accessKey_and_secretKey_tmp file
+  if [[ ! -f $MINIO_ACCESSKEY_SECRETKEY_TMP ]];then
+    echo "not find tmp file."
+    return 1
+  fi
+  if [[ $(cat $MINIO_ACCESSKEY_SECRETKEY_TMP|wc -l) -ne 2 ]];then
+    echo "tmp file is invalidate."
```
Comment (harshavardhana) [inline helm/minio/templates/_helper_create_user.txt]:
```suggestion
    echo "credentials file is invalid"
```

File: helm/minio/templates/_helper_create_user.txt [hunk 2]
```diff
@@ -35,28 +39,35 @@ connectToMinio() {
...
+# createUser ($policy)
 createUser() {
-  USER=$1
-  PASS=$2
-  POLICY=$3
-
+  POLICY=$1
+  #check accessKey_and_secretKey_tmp file
+  if [[ ! -f $MINIO_ACCESSKEY_SECRETKEY_TMP ]];then
+    echo "not find tmp file."
```
Comment (harshavardhana) [inline helm/minio/templates/_helper_create_user.txt]:
```suggestion
    echo "credentials file does not exist"
```

File: helm/minio/templates/_helper_create_user.txt [hunk 3]
```diff
@@ -35,28 +39,35 @@ connectToMinio() {
...
+    echo "tmp file is invalidate."
+    rm -f $MINIO_ACCESSKEY_SECRETKEY_TMP
+    return 1
+  fi
   # Create the user if it does not exist
-  if ! checkUserExists $USER ; then
+  if ! checkUserExists ; then
     echo "Creating user '$USER'"
-    ${MC} admin user add myminio $USER $PASS
+    ${MC} admin user add myminio $(head -1 $MINIO_ACCESSKEY_SECRETKEY_TMP) $(tail -1 $MINIO_ACCESSKEY_SECRETKEY_TMP)
```
Comment (harshavardhana) [inline helm/minio/templates/_helper_create_user.txt]:
```suggestion
    cat $MINIO_ACCESSKEY_SECRETKEY_TMP | ${MC} admin user add myminio
```

File: helm/minio/templates/_helper_create_user.txt [hunk 4]
```diff
@@ -79,10 +90,13 @@ connectToMinio $scheme
 {{ $global := . }}
 # Create the users
 {{- range .Values.users }}
+echo {{ tpl .accessKey $global }} > $MINIO_ACCESSKEY_SECRETKEY_TMP
 {{- if .existingSecret }}
-createUser {{ tpl .accessKey $global }} $(cat /config/secrets/{{ tpl .accessKey $global }}) {{ .policy }}
+cat /config/secrets/{{ tpl .accessKey $global }} >> $MINIO_ACCESSKEY_SECRETKEY_TMP
```
Comment (harshavardhana) [inline helm/minio/templates/_helper_create_user.txt]:
```suggestion
cat /config/secrets/{{ tpl .secretKey $global }} >> $MINIO_ACCESSKEY_SECRETKEY_TMP
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: helm/minio/templates/_helper_create_user.txt (hunk 3 above)
Comment (dormanze) [reply helm/minio/templates/_helper_create_user.txt]: This is a good idea.

File: helm/minio/templates/_helper_create_user.txt (hunk 1 above)
Comment (dormanze) [reply helm/minio/templates/_helper_create_user.txt]: This is a good suggestion.

File: helm/minio/templates/_helper_create_user.txt (hunk 4 above)
Comment (dormanze) [reply helm/minio/templates/_helper_create_user.txt]: This is a good suggestion.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-fs.sh | :x: [more...](#844ce654-74b9-472a-bc45-45fd0b07a40e) |
<a name="844ce654-74b9-472a-bc45-45fd0b07a40e"></a>
### 15243-5c1451f/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s7:32587
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 52 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 22 seconds
FAILED in 5 minutes and 10 seconds
Unhandled exception. System.AggregateException: One or more errors occurred. (A task was canceled.)
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Threading.Tasks.Task.GetExceptions(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.Program.Main(String[] args) in /mint/run/core/minio-dotnet/temp/Minio.Functional.Tests/Program.cs:line 156
--- End of stack trace from previous location ---

   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.Program.Main(String[] args) in /mint/run/core/minio-dotnet/temp/Minio.Functional.Tests/Program.cs:line 156
(8/15) Running minio-go tests ... done in 56 seconds
(9/15) Running minio-java tests ... done in 19 seconds
(10/15) Running minio-js tests ... done in 59 seconds
(11/15) Running minio-py tests ... done in 45 seconds
(12/15) Running s3cmd tests ... done in 42 seconds
(13/15) Running s3select tests ... done in 1 seconds
(14/15) Running versioning tests ... done in 3 minutes and 2 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Comment by sathieu** [comment]
@dormanze The chart is now broken with:

```yaml
  users:
  - accessKey: "foobar"
    existingSecret: minio-secretkeys-gitops
    existingSecretKey: "foobar"
    policy: "readwrite_test"
```

**Comment by sathieu** [comment]
Proposed fix: https://github.com/minio/minio/pull/15386

**Comment by dormanze** [comment]
> @dormanze The chart is now broken with:
> 
> ```yaml
>   users:
>   - accessKey: "foobar"
>     existingSecret: minio-secretkeys-gitops
>     existingSecretKey: "foobar"
>     policy: "readwrite_test"
> ```
LGTM

**Pull request merged**